src/metabase/driver/generic_sql/query_processor.clj

```diff
@@ -74,84 +72,83 @@
-  honeysql.types.SqlCall (formatted [this] this) ; already converted to HoneySQL so just return them as-is
-
-  Expression
-  (formatted [{:keys [operator args]}]
-    (apply (partial hsql/call operator)
-           (map formatted args)))
-
-  ExpressionRef
-  (formatted [{:keys [expression-name]}]
-    ;; Unfortunately you can't just refer to the expression by name in other clauses like filter, but have to use the
-    ;; original formuala.
-    (formatted (expression-with-name expression-name)))
-
-  Field
-  (formatted [{:keys [schema-name table-name special-type field-name]}]
-    (let [field (keyword (hx/qualify-and-escape-dots schema-name table-name field-name))]
-      (cond
-        (isa? special-type :type/UNIXTimestampSeconds)      (sql/unix-timestamp->timestamp (driver) field :seconds)
-        (isa? special-type :type/UNIXTimestampMilliseconds) (sql/unix-timestamp->timestamp (driver) field :milliseconds)
-        :else                                               field)))
-
-  FieldLiteral
-  (formatted [{:keys [field-name]}]
-    (keyword (hx/escape-dots (name field-name))))
-
-  DateTimeField
-  (formatted [{unit :unit, field :field}]
-    (sql/date (driver) unit (formatted field)))
-
-  BinnedField
-  (formatted [{:keys [bin-width min-value max-value field]}]
-    (let [formatted-field (formatted field)]
-      ;;
-      ;; Equation is | (value - min) |
-      ;;             | ------------- | * bin-width + min-value
-      ;;             |_  bin-width  _|
-      ;;
-      (-> formatted-field
-          (hx/- min-value)
-          (hx// bin-width)
-          hx/floor
-          (hx/* bin-width)
-          (hx/+ min-value))))
-
-  ;; e.g. the ["aggregation" 0] fields we allow in order-by
-  AgFieldRef
-  (formatted [{index :index}]
-    (let [{:keys [aggregation-type]} (aggregation-at-index index)]
-      ;; For some arcane reason we name the results of a distinct aggregation "count",
-      ;; everything else is named the same as the aggregation
-      (if (= aggregation-type :distinct)
-        :count
-        aggregation-type)))
-
-  Value
-  (formatted [value] (sql/prepare-value (driver) value))
-
-  DateTimeValue
-  (formatted [{{unit :unit} :field, :as value}]
-    (sql/date (driver) unit (sql/prepare-value (driver) value)))
-
-  RelativeDateTimeValue
-  (formatted [{:keys [amount unit], {field-unit :unit} :field}]
-    (sql/date (driver) field-unit (if (zero? amount)
-                                    (sql/current-datetime-fn (driver))
-                                    (driver/date-interval (driver) unit amount)))))
-
 
+(defmulti ^{:doc          (str "Return an appropriate HoneySQL form for an object. Dispatches off both driver and object "
+                               "classes making this easy to override in any places needed for a given driver.")
+            :arglists     '([driver x])
+            :style/indent 1}
+  ->honeysql
+  (fn [driver x]
+    [(class driver) (class x)]))
```

Comment:
I'm confused as to how any of the Object invocations below work. `(class driver)` will return the class of the driver, which is an object, but the defmethod stuff uses `=`, which doesn't know about inheritance. I thought maybe there was a `derive` or a `:default` sitting somewhere that would invoke it, but I don't see that either.

Reply:
@senior I'm pretty sure `defmethod` uses `isa?`. That's why you can use keywords and stuff with it if you `derive` them. For example what we do in the Druid QP https://github.com/metabase/metabase/blob/master/src/metabase/driver/druid/query_processor.clj#L426

Reply:
https://clojure.org/reference/multimethods

> Derivation is determined by a combination of either Java inheritance (for class values), or using Clojure’s ad hoc hierarchy system. The hierarchy system supports derivation relationships between names (either symbols or keywords), and relationships between classes and names. The derive function creates these relationships, and the isa? function tests for their existence. Note that isa? is not instance?.

Reply:
> Multimethods use isa? rather than = when testing for dispatch value matches. Note that the first test of isa? is =, so exact matches work.

Reply:
👍 not sure why that didn't work for me at the repl earlier, just tried it again and it works fine